File: graphlite-sdk/examples/basic_usage.rs
//! Basic usage example for GraphLite SDK
//!
//! This example demonstrates the core features of the GraphLite Rust SDK:
//! - Opening a database
//! - Creating sessions
//! - Executing queries
//! - Using transactions
//! - Query builder API
//! - Typed result deserialization
//!
//! Run with: cargo run --example basic_usage

use graphlite_sdk::{Error, GraphLite};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct Person {
    name: String,
    age: f64,
}

fn main() -> Result<(), Error> {
    println!("=== GraphLite SDK Basic Usage Example ===\n");

    // 1. Open a database
    println!("1. Opening database...");
    let db_path = "/tmp/graphlite_sdk_example";
    let db = GraphLite::open(db_path)?;
    println!("   ✓ Database opened at {}\n", db_path);

    // 2. Create a session
    println!("2. Creating session...");
    let session = db.session("admin")?;
    println!("   ✓ Session created for user 'admin'\n");

    // 3. Execute DDL statements
    println!("3. Creating schema and graph...");
    session.execute("CREATE SCHEMA IF NOT EXISTS example")?;
    session.execute("USE SCHEMA example")?;
    session.execute("CREATE GRAPH IF NOT EXISTS social")?;
    session.execute("USE GRAPH social")?;
    println!("   ✓ Schema and graph created\n");

    // 4. Insert data using transactions
    println!("4. Inserting data with transaction...");
    {
        let mut tx = session.transaction()?;
        tx.execute("CREATE (p:Person {name: 'Alice', age: 30})")?;
        tx.execute("CREATE (p:Person {name: 'Bob', age: 25})")?;
        tx.execute("CREATE (p:Person {name: 'Charlie', age: 35})")?;
        tx.commit()?;
        println!("   ✓ Inserted 3 persons\n");
    }

    // 5. Query data directly
    println!("5. Querying data...");
    let result = session.query("MATCH (p:Person) RETURN p.name as name, p.age as age")?;
    println!("   Found {} persons:", result.rows.len());
    for row in &result.rows {
        if let (Some(name), Some(age)) = (row.get_value("name"), row.get_value("age")) {
            println!("   - Name: {:?}, Age: {:?}", name, age);
        }
    }
    println!();

    // 6. Use query builder
    println!("6. Using query builder...");
    let result = session
        .query_builder()
        .match_pattern("(p:Person)")
        .where_clause("p.age > 25")
        .return_clause("p.name as name, p.age as age")
        .order_by("p.age DESC")
        .execute()?;
    println!("   Found {} persons over 25:", result.rows.len());
    for row in &result.rows {
        if let (Some(name), Some(age)) = (row.get_value("name"), row.get_value("age")) {
            println!("   - Name: {:?}, Age: {:?}", name, age);
        }
    }
    println!();

    // 7. Typed deserialization
    println!("7. Using typed deserialization...");
    let result = session.query("MATCH (p:Person) RETURN p.name as name, p.age as age")?;
    let typed = graphlite_sdk::TypedResult::from(result);
    let people: Vec<Person> = typed.deserialize_rows()?;
    println!("   Deserialized {} persons:", people.len());
    for person in &people {
        println!("   - {:?}", person);
    }
    println!();

    // 8. Transaction with rollback
    println!("8. Demonstrating transaction rollback...");
    {
        let mut tx = session.transaction()?;
        tx.execute("CREATE (p:Person {name: 'David', age: 40})")?;
        println!("   Created person 'David' in transaction");
        // Transaction is dropped without commit - automatically rolls back
    }
    println!("   Transaction rolled back (David not persisted)\n");

    // 9. Verify rollback
    let result = session.query("MATCH (p:Person) RETURN count(p) as count")?;
    if let Some(row) = result.rows.first() {
        if let Some(count) = row.get_value("count") {
            println!("   Person count after rollback: {:?}\n", count);
        }
    }

    println!("=== Example completed successfully ===");
    Ok(())
}

File: sdk-python/examples/basic_usage.py
"""
GraphLite Python SDK - Basic Usage Example

This example demonstrates the core features of the GraphLite Python SDK:
- Opening a database
- Creating sessions
- Executing queries
- Using transactions
- Query builder API
- Typed result deserialization

Run with: python3 examples/basic_usage.py
"""

from dataclasses import dataclass
from typing import Optional
import tempfile
import shutil
import sys
from pathlib import Path

# Import from the installed package
from graphlite_sdk import (
    GraphLite,
    GraphLiteError,
    SerializationError,
    QueryError,
    TransactionError,
    TypedResult,
)

@dataclass
class Person:
    """Person entity for typed deserialization"""
    name: str
    age: int


def main() -> int:
    """Run the basic usage example"""
    print("=== GraphLite SDK Basic Usage Example ===\n")

    # Use temporary directory for demo
    db_path = tempfile.mkdtemp(prefix="graphlite_sdk_example_")
    
    try:
        # 1. Open a database
        print("1. Opening database...")
        db = GraphLite.open(db_path)
        print(f"   Database opened at {db_path}\n")

        # 2. Create a session
        print("2. Creating session...")
        session = db.session("admin")
        print("   Session created for user 'admin'\n")

        # 3. Execute DDL statements
        print("3. Creating schema and graph...")
        session.execute("CREATE SCHEMA IF NOT EXISTS /example")
        session.execute("SESSION SET SCHEMA /example")
        session.execute("CREATE GRAPH IF NOT EXISTS social")
        session.execute("SESSION SET GRAPH social")
        print("   Schema and graph created\n")

        # 4. Insert data using transactions
        print("4. Inserting data with transaction...")
        with session.transaction() as tx:
            tx.execute("INSERT (p:Person {name: 'Alice', age: 30})")
            tx.execute("INSERT (p:Person {name: 'Bob', age: 25})")
            tx.execute("INSERT (p:Person {name: 'Charlie', age: 35})")
            tx.execute("INSERT (p:Person {name: 'David', age: 28})")
            tx.execute("INSERT (p:Person {name: 'Eve', age: 23})")
            tx.execute("INSERT (p:Person {name: 'Frank', age: 40})")
            tx.commit()
        print("   Inserted 6 persons\n")

        # 5. Query data directly
        print("5. Querying data...")
        result = session.query("MATCH (p:Person) RETURN p.name as name, p.age as age")
        print(f"   Found {len(result.rows)} persons:")
        for row in result.rows:
            name = row.get("name")
            age = row.get("age")
            if name is not None and age is not None:
                print(f"   - Name: {name}, Age: {age}")
        print()

        # 6. Use query builder
        print("6. Using query builder...")
        result = (session.query_builder()
            .match_pattern("(p:Person)")
            .where_clause("p.age > 25")
            .return_clause("p.name as name, p.age as age")
            .order_by("p.age DESC")
            .execute())
        print(f"   Found {len(result.rows)} persons over 25:")
        for row in result.rows:
            name = row.get("name")
            age = row.get("age")
            if name is not None and age is not None:
                print(f"   - Name: {name}, Age: {age}")
        print()

        # 7. Typed deserialization
        print("7. Using typed deserialization...")
        result = session.query("MATCH (p:Person) RETURN p.name as name, p.age as age")
        typed = TypedResult(result)
        people = typed.deserialize_rows(Person)
        print(f"   Deserialized {len(people)} persons:")
        for person in people:
            print(f"   - {person}")
        print()

        # 8. Transaction with rollback
        print("8. Demonstrating transaction rollback...")
        try:
            with session.transaction() as tx:
                tx.execute("INSERT (p:Person {name: 'George', age: 50})")
                print("   Inserted person 'George' in transaction")
                # Transaction is NOT committed - will auto-rollback
                # (by not calling tx.commit())
        except Exception:
            pass  # Expected - rollback on exception
        print("   Transaction rolled back (George not persisted)\n")

        # 9. Verify rollback
        result = session.query("MATCH (p:Person) RETURN count(p) as count")
        if result.rows:
            count = result.rows[0].get("count")
        print(f"   Person count after rollback: {count}\n")
        print("=== Example completed successfully ===")
        return 0

    except GraphLiteError as e:
        print(f"\n[ERROR] GraphLite Error: {e}")
        return 1

    except Exception as e:
        print(f"\n[ERROR] Unexpected error: {e}")
        import traceback
        traceback.print_exc()
        return 1

    finally:
        # Cleanup temporary directory
        shutil.rmtree(db_path, ignore_errors=True)


if __name__ == "__main__":
    exit(main())
